docs/source/runtime_troubles.rst

```diff
@@ -0,0 +1,311 @@
+.. _runtime_troubles:
+
+Annotation issues at runtime
+============================
+
+Idiomatic use of type annotations can sometimes run up against what a given
+version of Python considers legal code. This section describes these scenarios
+and explains how to get your code running again. Generally speaking, we have
+three tools at our disposal:
+
+* For Python 3.7 through 3.9, use of ``from __future__ import annotations``
+  (:pep:`563`), made the default in Python 3.10 and later
+* Use of string literal types or type comments
+* Use of ``typing.TYPE_CHECKING``
+
+We provide a description of these before moving onto discussion of specific
+problems you may encounter.
+
+.. _string-literal-types:
+
+String literal types
+--------------------
+
+Type comments can't cause runtime errors because comments are not evaluated by
+Python. In a similar way, using string literal types sidesteps the problem of
+annotations that would cause runtime errors.
+
+Any type can be entered as a string literal, and you can combine
+string-literal types with non-string-literal types freely:
+
+.. code-block:: python
+
+   def f(a: List['A']) -> None: ...  # OK
+   def g(n: 'int') -> None: ...      # OK, though not useful
+
+   class A: pass
+
+String literal types are never needed in ``# type:`` comments and :ref:`stub files <stub-files>`.
+
+String literal types must be defined (or imported) later *in the same module*.
+They cannot be used to leave cross-module references unresolved.  (For dealing
+with import cycles, see :ref:`import-cycles`.)
+
+.. _future-annotations:
+
+Future annotations import (PEP 563)
+-----------------------------------
+
+Many of the issues described here are caused by Python trying to evaluate
+annotations. From Python 3.10 on, Python will no longer attempt to evaluate
+function and variable annotations. This behaviour is made available in Python
+3.7 and later through the use of ``from __future__ import annotations``.
+
+This can be thought of as automatic string literal-ification of all function and
+variable annotations. Note that function and variable annotations are still
+required to be valid Python syntax. For more details, see :pep:`563`.
+
+.. note::
+
+    Even with the ``__future__`` import, there are some scenarios that could
+    still require string literals or result in errors, typically involving use
+    of forward references or generics in:
+
+    * :ref:`type aliases <type-aliases>`;
+    * :ref:`casts <casts>`;
+    * type definitions (see :py:class:`~typing.TypeVar`, :py:func:`~typing.NewType`, :py:class:`~typing.NamedTuple`);
+    * base classes.
+
+    .. code-block:: python
+
+        # base class example
```

Comment:
Maybe add `from __future__ import annotations` to this example to make it explicit that string literals are still needed?